src/lem/circuit.rs

```diff
@@ -0,0 +1,889 @@
+//! LEM (for the third time), adding implication constraints, glueing together all the
+//! information. This step is better explained in the `synthesize` function.
+
+use std::collections::{HashMap, HashSet};
+
+use anyhow::{anyhow, bail, Context, Result};
+use bellperson::{
+    gadgets::{boolean::Boolean, num::AllocatedNum},
+    ConstraintSystem,
+};
+use indexmap::{IndexMap, IndexSet};
+
+use crate::circuit::gadgets::{
+    constraints::{
+        alloc_equal_const, and, enforce_selector_with_premise, implies_equal, implies_equal_zero,
+    },
+    data::hash_poseidon,
+    pointer::AllocatedPtr,
+};
+
+use crate::field::{FWrap, LurkField};
+
+use super::{
+    interpreter::{Frame, SlotType},
+    path::Path,
+    pointers::ZPtr,
+    store::Store,
+    MetaPtr, LEM, LEMOP,
+};
+
+/// Holds a counter per path. We want to use this to count the number of slots
+/// that have already been used on each LEM path.
+#[derive(Default)]
+struct PathTicker(HashMap<Path, usize>);
+
+impl PathTicker {
+    /// Increments the counter of a path. If the path wasn't tracked, returns
+    /// `0` and starts tracking it
+    pub(crate) fn tick(&mut self, path: Path) -> usize {
+        let next = self.0.get(&path).unwrap_or(&0).to_owned();
+        self.0.insert(path, next + 1);
+        next
+    }
+
+    /// Starts tracking a new path with a counter from another. If the reference
+    /// path wasn't being tracked, the new one won't be either, such that calling
+    /// `next` will return `0`.
+    pub(crate) fn cont(&mut self, new: Path, from: &Path) {
+        match self.0.get(from) {
+            Some(i) => {
+                self.0.insert(new, *i);
+            }
+            None => (),
+        }
+    }
+}
+
+/// Keeps track of previously used slots indices for each possible LEM path,
+/// being capable of informing the next slot index to be used
+#[derive(Default)]
+struct SlotsCounter {
+    hash2: PathTicker,
+    hash3: PathTicker,
+    hash4: PathTicker,
+}
+
+impl SlotsCounter {
+    #[inline]
+    pub(crate) fn next_hash2(&mut self, path: Path) -> usize {
+        self.hash2.tick(path)
+    }
+
+    #[inline]
+    pub(crate) fn next_hash3(&mut self, path: Path) -> usize {
+        self.hash3.tick(path)
+    }
+
+    #[inline]
+    pub(crate) fn next_hash4(&mut self, path: Path) -> usize {
+        self.hash4.tick(path)
+    }
+
+    pub(crate) fn cont(&mut self, new: Path, from: &Path) {
+        self.hash2.cont(new.clone(), from);
+        self.hash3.cont(new.clone(), from);
+        self.hash4.cont(new, from);
+    }
+}
+
+/// Contains a `slots_map` that maps `LEMOP`s to their slots. This map is not
+/// expected to be injective, as two or more `LEMOP`s can be mapped to the same
+/// slot. The `slots` attribute is an set of all slots present on `slots_map`.
+pub struct SlotsInfo {
+    slots_map: IndexMap<LEMOP, (usize, SlotType)>,
+    slots: IndexSet<(usize, SlotType)>,
+}
```

Comment:
What is the `usize`? That is, what does a `(usize, SlotType)` pair represent? I know it's a 'slot', but the answer is not obvious.

Are all of these 'slots' indexed? If so, maybe it would be clearer (certainly more 'normalized') to just store the index in the map.

Reply:
I added this in the interpreter:
```
/// This hashmap is populated during interpretation, telling which slots were
/// visited and the pointers that were collected for each of them.
///
/// `(usize, SlotType)` is a pair of slot index and slot type, as shown below:
///
///```text
///            Slot index
///      ┌────┬───┬───┬───┬───
///      │    │ 0 │ 1 │ 2 │...
///      ├────┼───┼───┼───┼───
/// Slot │ H2 │ a │ b │   │...
/// type ├────┼───┼───┼───┼───
///      │ H3 │   │ c │ d │...
///      ├────┼───┼───┼───┼───
///      │... │...│...│...│...
///```
///
/// In the example above, we can see three visited slots:
/// * The slots indices 0 and 1 for `Hash2`
/// * The slots indices 1 and 2 for `Hash3`
pub(crate) type Visits<F> = HashMap<(usize, SlotType), Vec<Ptr<F>>>;
```
Maybe that explanation is in the wrong place. Or maybe there should be a `Slot` type and the code should use it elsewhere

Reply:
That example is not accurate. I'll fix it

Reply:
In general, leaning on the type system to help document meaning will be extremely helpful. A lot of review time on this work to date has been spent just trying to untangle what various unlabeled tuples are meant to represent.

Reply:
Maybe it's fine — just trying to get a handle on what the actual structure of the data even is. I think I get it now.

Reply:
I don't understand it as an answer to my question here, so maybe something does need to be here (but not necessarily that). I don't know what the table is showing me because I don't know what a, b, c, d represent.

Reply:
A "slot" is characterized by an index and a slot type
"the first slot for Hash2, the first slot for Hash3 etc"

Reply:
We now have a `Slot` struct for clarity